File: csharp/10-null-spiral/test/room_lookup_test.cs
using Xunit;

class FakeDb : IRoomDb
{
    private string? _missing;

    public FakeDb(string? missing = null)
    {
        _missing = missing;
    }

    public Student? GetStudent(int id) => _missing == "student" ? null : new Student(1);
    public Enrollment? GetEnrollment(int id) => _missing == "enrollment" ? null : new Enrollment(2);
    public Course? GetCourse(int id) => _missing == "course" ? null : new Course(3);
    public Section? GetSection(int id) => _missing == "section" ? null : new Section(4);
    public Room? GetRoom(int id) => _missing == "room" ? null : new Room("Room A");
}

public class RoomLookupTest
{
    [Fact]
    public void ReturnsRoomNameForValidChain()
    {
        var lookup = new RoomLookup();
        var result = lookup.get_room_for_student(1, new FakeDb());
        Assert.Equal("Room A", result);
    }

    [Fact]
    public void ReturnsNullForNullDb()
    {
        var lookup = new RoomLookup();
        var result = lookup.get_room_for_student(1, null);
        Assert.Null(result);
    }

    [Fact]
    public void ReturnsNullForNullStudent()
    {
        var lookup = new RoomLookup();
        var result = lookup.get_room_for_student(1, new FakeDb("student"));
        Assert.Null(result);
    }

    [Fact]
    public void ReturnsNullForNullEnrollment()
    {
        var lookup = new RoomLookup();
        var result = lookup.get_room_for_student(1, new FakeDb("enrollment"));
        Assert.Null(result);
    }

    [Fact]
    public void ReturnsNullForNullCourse()
    {
        var lookup = new RoomLookup();
        var result = lookup.get_room_for_student(1, new FakeDb("course"));
        Assert.Null(result);
    }

    [Fact]
    public void ReturnsNullForNullSection()
    {
        var lookup = new RoomLookup();
        var result = lookup.get_room_for_student(1, new FakeDb("section"));
        Assert.Null(result);
    }

    [Fact]
    public void ReturnsNullForNullRoom()
    {
        var lookup = new RoomLookup();
        var result = lookup.get_room_for_student(1, new FakeDb("room"));
        Assert.Null(result);
    }
}

File: golang/10-null-spiral/room_lookup_test.go
package kata

import "testing"

type fakeDB struct{ missing string }

func (d fakeDB) GetStudent(int) *Student {
	if d.missing == "student" {
		return nil
	}
	return &Student{1}
}
func (d fakeDB) GetEnrollment(int) *Enrollment {
	if d.missing == "enrollment" {
		return nil
	}
	return &Enrollment{2}
}
func (d fakeDB) GetCourse(int) *Course {
	if d.missing == "course" {
		return nil
	}
	return &Course{3}
}
func (d fakeDB) GetSection(int) *Section {
	if d.missing == "section" {
		return nil
	}
	return &Section{4}
}
func (d fakeDB) GetRoom(int) *Room {
	if d.missing == "room" {
		return nil
	}
	return &Room{"Room A"}
}

func TestGetRoomForStudentReturnsRoomNameForValidChain(t *testing.T) {
	rl := NewRoomLookup()
	result := rl.get_room_for_student(1, fakeDB{})
	if result == nil {
		t.Fatal("expected non-nil result")
	}
	if *result != "Room A" {
		t.Errorf("expected 'Room A', got %q", *result)
	}
}

func TestGetRoomForStudentReturnsNullForNullDb(t *testing.T) {
	rl := NewRoomLookup()
	result := rl.get_room_for_student(1, nil)
	if result != nil {
		t.Errorf("expected nil result, got %v", result)
	}
}

func TestGetRoomForStudentReturnsNullForNullStudent(t *testing.T) {
	rl := NewRoomLookup()
	result := rl.get_room_for_student(1, fakeDB{missing: "student"})
	if result != nil {
		t.Errorf("expected nil result, got %v", result)
	}
}

func TestGetRoomForStudentReturnsNullForNullEnrollment(t *testing.T) {
	rl := NewRoomLookup()
	result := rl.get_room_for_student(1, fakeDB{missing: "enrollment"})
	if result != nil {
		t.Errorf("expected nil result, got %v", result)
	}
}

func TestGetRoomForStudentReturnsNullForNullCourse(t *testing.T) {
	rl := NewRoomLookup()
	result := rl.get_room_for_student(1, fakeDB{missing: "course"})
	if result != nil {
		t.Errorf("expected nil result, got %v", result)
	}
}

func TestGetRoomForStudentReturnsNullForNullSection(t *testing.T) {
	rl := NewRoomLookup()
	result := rl.get_room_for_student(1, fakeDB{missing: "section"})
	if result != nil {
		t.Errorf("expected nil result, got %v", result)
	}
}

func TestGetRoomForStudentReturnsNullForNullRoom(t *testing.T) {
	rl := NewRoomLookup()
	result := rl.get_room_for_student(1, fakeDB{missing: "room"})
	if result != nil {
		t.Errorf("expected nil result, got %v", result)
	}
}
